types/underscore/underscore-tests.ts

```diff
@@ -1116,6 +1047,142 @@ declare const extractChainTypes: ChainTypeExtractor;
     _.chain([{a: 'a'}, {a: 'b'}]).findLastIndex({ a: 'b' }).value(); // $ExpectType number
 }
 
+// Objects
+
+// isEqual
+{
+    _.isEqual(anyValue, anyValue); // $ExpectType boolean
+    _(anyValue).isEqual(anyValue); // $ExpectType boolean
+    extractChainTypes(_.chain(anyValue).isEqual(anyValue)); // $ExpectType ChainType<boolean, never>
+}
+
+// isEmpty
+{
+    _.isEmpty(anyValue); // $ExpectType boolean
+    _(anyValue).isEmpty(); // $ExpectType boolean
+    extractChainTypes(_.chain(anyValue).isEmpty()); // $ExpectType ChainType<boolean, never>
+}
+
+// isMatch
+{
+    _.isMatch(anyValue, anyValue); // $ExpectType boolean
+    _(anyValue).isMatch(anyValue); // $ExpectType boolean
+    extractChainTypes(_.chain(anyValue).isMatch(anyValue)); // $ExpectType ChainType<boolean, never>
+}
+
+// isElement
+{
+    _.isElement(anyValue) ? anyValue : neverValue; // $ExpectType Element
```

Comment:
Please explain to me how this works. This reminds me a bit of undefined behavior optimizations in C and C++ compilers, which is not a positive association in this context...

Reply:
The use of `_.isElement` in the check part of the ternary statement modifies the type of `anyValue` on the truthy branch of the statement to `Element`, TS determines the result type by unioning the types of the truthy and falsy branches of the ternary statement to `Element | never`, and the type is simplified to `Element` because [never is a subtype of every type](https://www.typescriptlang.org/docs/handbook/release-notes/overview.html#the-never-type).

I'd be open to changing this to something else if you'd prefer. This seemed like a not-totally-unreasonable concise way to get the modified truthy type out of these checks, but the `never` magic to deal with the existence of the falsy branch is admittedly kind of hand-wavey.

Reply:
Well, "magic" was probably not the best word to use there since this behavior is at least explicitly called out in the documentation for the release of the `never` type, which says "Because never is a subtype of every type, it is always omitted from union types."

Reply:
Given this explanation, I'm perfectly happy with the approach you've taken. A union with an identity type is fine.

I was a bit afraid that this was some kind of fuzzy logic thing where the TS compiler goes "oh, the true branch returns `never`. That cannot be the case so I guess we'll take the other branch".